pose_lookup/plugin.cpp

```diff
@@ -0,0 +1,172 @@
+#include "common/phonebook.hpp"
+#include "common/pose_prediction.hpp"
+#include "common/data_format.hpp"
+#include "common/plugin.hpp"
+
+/*pyh: reusing data_loading from ground_truth_slam*/
+#include "data_loading.hpp"
+
+using namespace ILLIXR;
+
+class pose_lookup_impl : public pose_prediction {
+public:
+    pose_lookup_impl(const phonebook* const pb)
+		: sb{pb->lookup_impl<switchboard>()}
+	, _m_sensor_data{load_data()}
+	, _m_sensor_data_it{_m_sensor_data.cbegin()}
+	, dataset_first_time{_m_sensor_data_it->first}
+	, _m_start_of_time{std::chrono::high_resolution_clock::now()}
+	, _m_vsync_estimate{sb->subscribe_latest<time_type>("vsync_estimate")}
+    {
+    	auto newoffset = correct_pose(_m_sensor_data_it->second).orientation;
+    	set_offset(newoffset);
+    }
+
+    virtual fast_pose_type get_fast_pose() override {
+	return get_fast_pose( std::chrono::system_clock::now() );
+    }
+
+    virtual pose_type get_true_pose() override {
+	throw std::logic_error{"Not Implemented"};
+    }
+
+
+    virtual bool fast_pose_reliable() {
+	return true;
+    }
+
+    virtual bool true_pose_reliable() {
+	return false;
+    }
+
+   virtual void set_offset(const Eigen::Quaternionf& raw_o_times_offset) override{
+	std::lock_guard<std::mutex> lock {offset_mutex};
+	Eigen::Quaternionf raw_o = raw_o_times_offset * offset.inverse();
+	//std::cout << "pose_prediction: set_offset" << std::endl;
+	offset = raw_o.inverse();
+    }
+
+    Eigen::Quaternionf apply_offset(const Eigen::Quaternionf& orientation) const {
+	std::lock_guard<std::mutex> lock {offset_mutex};
+	return orientation * offset;
+    }
+    virtual fast_pose_type get_fast_pose(time_type time) override {
+		/*
+			ullong vsync_time = std::chrono::nanoseconds(get_vsync(time).time_since_epoch()).count();
+			ullong plug_in_time = std::chrono::nanoseconds(_m_start_of_time.time_since_epoch()).count();
+
+			std::cout<<"dataset_first: "<<dataset_first_time<<"vsync time: "<<vsync_time <<" plug_in_time: "<<plug_in_time<<std::endl;
+			std::cout<<"lookup time: "<<lookup_time <<std::endl;
+			std::cout<<"data end time: "<< _m_sensor_data.rbegin()->first <<std::endl;
+		*/
+	    	//set_offset(correct_pose(_m_sensor_data.begin()->second).orientation);
+
+		const time_type* estimated_vsync = _m_vsync_estimate->get_latest_ro();
+		time_type vsync;
+		if(estimated_vsync == nullptr) {
+			vsync = std::chrono::system_clock::now();
```

Comment:
Is making vsync now if it is null the right assumption? Should it be logging an error instead?

Reply:
Let's log to stderr in addition. This condition would be true be if a plugin asks for pose _before the first vsync_. In such a case, getting a pose for "now" is an acceptable approximation.